### ISSUE TITLE
[Data] Cannot Sort without Key

### ISSUE THREAD

**Issue by ivanthewebber** [ray-project/ray#49198]
### What happened + What you expected to happen

`Dataset.sort`'s type hint for the `key` parameter is `Union[str, List[str], None] = None`. It would be expected that if `key` is none the rows are sorted on all columns. However, providing `None` causes an error due to unguarded indexing in [SortKey](https://github.com/ray-project/ray/blob/23ef2aa00d52c7d4bc65c87d09214754707d043e/python/ray/data/_internal/planner/exchange/sort_task_spec.py#L20) (ray.data._internal.planner.exchange.sort_task_spec.py.SortKey).

These lines set `descending` which will be `[]` if `key` is `None`.
```python
        if key is None:
            key = []
        if isinstance(key, str):
            key = [key]
        if not (isinstance(key, list) and all(isinstance(k, str) for k in key)):
            raise ValueError(
                f"Key must be a string or a list of strings, but got {key}."
            )
        if isinstance(descending, bool):
            descending = [descending for _ in key]
```
```sh
...
```
Farther down the file descending is accessed without a size check.
```python
    def get_descending(self) -> bool:
        return self._descending[0]

    def to_arrow_sort_args(self) -> List[Tuple[str, str]]:
        return [
            (key, "descending" if self._descending[0] else "ascending")
            for key in self._columns
        ]

    def to_pandas_sort_args(self) -> Tuple[List[str], bool]:
        return self._columns, not self._descending[0]
```

The easy fix is to change [line 38](https://github.com/ray-project/ray/blob/23ef2aa00d52c7d4bc65c87d09214754707d043e/python/ray/data/_internal/planner/exchange/sort_task_spec.py#L38C13-L38C51) to `descending = [descending for _ in key] if key else [descending]`

### Versions / Dependencies

Ray 2.40

### Reproduction script

```python
import ray.data
ray.data.from_items([4, 1, 2, 2, 3, 3, 3]).sort(None).show()
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by wingkitlee0** [comment]
`None` was not working before. See the PR that removes it. https://github.com/ray-project/ray/pull/48969

**Comment by ivanthewebber** [comment]
It's also worth noting that `ray.data.Dataset.columns` isn't really `O(1)` like the docstring says for unmaterialized datasets. If you have any sort or groupby it will be really expensive. Thus users are forced to track their columns themselves or face poor performance.

I tried fixing it in a new branch and ran into some additional complexity because empty-columns is used as a signal for aggregations that should be done in one block.

**Comment by richardliaw** [comment]
Hmm, if we already merged #48969, this shouldn't occur anymore. Could you share your use case (why do you want to sort by none in the first place)?

**Comment by richardliaw** [comment]
(closing this issue for now, but if you have a use case let's discuss!)
